src/ir/access_path_selectors_set.h

```diff
@@ -0,0 +1,91 @@
+#ifndef SRC_IR_ACCESS_PATH_SELECTORS_SET_H_
+#define SRC_IR_ACCESS_PATH_SELECTORS_SET_H_
+
+#include "access_path_selectors.h"
+
+#include "absl/container/flat_hash_set.h"
+
+namespace raksha::ir {
+
+// This class implements a set of AccessPathSelectors. It does not actually
+// contain a set; it contains a vector. This makes mutating the "set" much
+// easier, as we add parent paths on as this is passed upwards. We expect
+// that invariants on the type tree will make most additions to the set
+// unique anyway. We lazily unique the entries when it matters.
+class AccessPathSelectorsSet {
+ public:
+  // The default constructor will construct an empty set.
+  AccessPathSelectorsSet() = default;
+
+  // Allow creating an AccessPathSelectorSet with an explicit list of items
+  // that should be in that set.
+  explicit AccessPathSelectorsSet(std::vector<AccessPathSelectors> contents)
+    : inner_vec_(std::move(contents)) {}
+
+  // Returns a set that is the union of the two passed-in sets.
+  static AccessPathSelectorsSet Union(
+      AccessPathSelectorsSet set1, AccessPathSelectorsSet set2) {
+    set1.inner_vec_.insert(
+        set1.inner_vec_.end(),
+        std::make_move_iterator(set2.inner_vec_.begin()),
+        std::make_move_iterator(set2.inner_vec_.end()));
+    return set1;
+  }
+
+  // Returns a set that is the intersection of the two passed-in sets.
+  static AccessPathSelectorsSet Intersect(
+      AccessPathSelectorsSet set1, AccessPathSelectorsSet set2) {
+    AccessPathSelectorsSet result;
+
+    // Make the first set into a hash set for efficient lookup.
+    absl::flat_hash_set<AccessPathSelectors> hash_set1 =
+      AccessPathSelectorsSet::CreateAbslSet(std::move(set1));
+
+    // Place items from the second set into the result only if they were in
+    // the first set.
+    for (AccessPathSelectors &path : set2.inner_vec_) {
+      if (hash_set1.contains(path)) {
+        result.inner_vec_.push_back(std::move(path));
+      }
+    }
+
+    return result;
+  }
+
+  // Returns a set that has the same elements as child_set but with
+  // parent_selector added as a parent to each of them.
+  static AccessPathSelectorsSet AddParentToAll(
+      Selector parent_selector, AccessPathSelectorsSet child_set) {
```

Comment:
I still am not getting used to the missing `const &` in the arguments. I believe we are making a copy of both the arguments in the function body?

Reply:
I forgot to use `std::move` on `parent_selector`, so I should do that. I think I'm doing the right thing on the loop over the `child_set` elements by moving each element temporary.

Reply:
Actually, I don't think moving `parent_selector` is the right thing. I'm putting `parent_selector` as the parent of each element of `child_set`; therefore, if `child_set` has N elements, I *do* need to make at least `N-1` copies of `parent_selector`. Therefore, `std::move` is probably not the right thing here, as we need the name to be attached to its value through each iteration of the loop.